packages/hardhat-core/src/builtin-tasks/compile.ts

```diff
@@ -161,9 +161,22 @@ subtask(TASK_COMPILE_SOLIDITY_READ_FILE)
   .addParam("absolutePath", undefined, undefined, types.string)
   .setAction(
     async ({ absolutePath }: { absolutePath: string }): Promise<string> => {
-      return fsExtra.readFile(absolutePath, {
-        encoding: "utf8",
-      });
+      try {
+        return await fsExtra.readFile(absolutePath, {
+          encoding: "utf8",
+        });
+      } catch (e) {
+        if (!fsExtra.existsSync(absolutePath)) {
+          throw new HardhatError(ERRORS.RESOLVER.FILE_NOT_FOUND);
+        }
+
+        if (fsExtra.lstatSync(absolutePath).isDirectory()) {
+          throw new HardhatError(ERRORS.RESOLVER.INVALID_IMPORT_OF_DIRECTORY);
```

Comment:
Sorry, I think I didn't explain this correctly in our call. We shouldn't throw resolver-related errors here, because this subtask is just about reading files. We should throw a more general "tried to read a file, but found a directory" error.

To be explicit, this is what I would do:
- Add a new error in the `GENERAL` category about trying to read a directory as a file
- In the resolver, use `isHardhatErrorType` to detect those kind of errors, and re-throw the `RESOLVER.INVALID_IMPORT_OF_DIRECTORY` one.
- Remove the previous `if` statement that also throws a resolver error. That was (I guess) handled correctly before, and so the `throw e` at the end should keep whatever behavior we have now.